internal/codespaces/rpc/invoker_test.go

```diff
@@ -3,84 +3,156 @@ package rpc
+		return &codespace.NotifyCodespaceOfClientActivityResponse{
+			Message: "",
+			Result:  true,
+		}, nil
+	}
 
-	// Stop the gRPC server when the test is done
-	t.Cleanup(func() {
-		cancel()
-	})
+	return server
 }
 
-func createTestInvoker(t *testing.T) Invoker {
-	t.Helper()
-
-	// Clear the stored client activity
-	rpctest.NotifyReceivedActivity = ""
-
-	invoker, err := CreateInvoker(context.Background(), &rpctest.Session{})
+func runTestServer(ctx context.Context, server *mockServer) error {
+	listener, err := net.Listen("tcp", "127.0.0.1:16634")
 	if err != nil {
-		t.Fatalf("error connecting to internal server: %v", err)
+		return fmt.Errorf("failed to listen: %w", err)
 	}
+	defer listener.Close()
+
+	s := grpc.NewServer()
+	jupyter.RegisterJupyterServerHostServer(s, server)
+	codespace.RegisterCodespaceHostServer(s, server)
+	ssh.RegisterSshServerHostServer(s, server)
 
-	t.Cleanup(func() {
-		testNotifyCodespaceOfClientActivity(t)
-		invoker.Close()
-	})
+	ch := make(chan error, 1)
+	go func() { ch <- s.Serve(listener) }()
 
-	return invoker
+	select {
+	case <-ctx.Done():
+		s.Stop()
+		<-ch
+		return nil
+	case err := <-ch:
+		return err
+	}
+}
+
+func waitAndReportError(t *testing.T, wg *errgroup.Group) {
+	if err := wg.Wait(); err != nil {
+		t.Fatalf("error reported from errgroup: %v", err)
+	}
 }
 
 // Test that the RPC invoker notifies the codespace of client activity on connection
-func testNotifyCodespaceOfClientActivity(t *testing.T) {
-	if rpctest.NotifyReceivedActivity != connectedEventName {
-		t.Fatalf("expected %s, got %s", connectedEventName, rpctest.NotifyMessage)
+func verifyNotifyCodespaceOfClientActivity(t *testing.T, server *mockServer) {
+	calls := server.CodespaceHostServerMock.NotifyCodespaceOfClientActivityCalls()
+	if len(calls) == 0 {
+		t.Fatalf("no client activity calls")
 	}
+
+	for _, call := range calls {
+		activities := call.NotifyCodespaceOfClientActivityRequest.GetClientActivities()
+		if activities[0] == connectedEventName {
+			return
+		}
+	}
+
+	t.Fatalf("no activity named %s", connectedEventName)
 }
 
 // Test that the RPC invoker returns the correct port and URL when the JupyterLab server starts successfully
 func TestStartJupyterServerSuccess(t *testing.T) {
-	startServer(t)
-	invoker := createTestInvoker(t)
+	resp := jupyter.GetRunningServerResponse{
+		Port:      strconv.Itoa(1234),
+		ServerUrl: "http://localhost:1234?token=1234",
+		Message:   "",
+		Result:    true,
+	}
+
+	server := newMockServer()
+	server.JupyterServerHostServerMock.GetRunningServerFunc = func(context.Context, *jupyter.GetRunningServerRequest) (*jupyter.GetRunningServerResponse, error) {
+		return &resp, nil
+	}
+
+	var wg errgroup.Group
+	defer waitAndReportError(t, &wg)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	wg.Go(func() error { return runTestServer(ctx, server) })
```

Comment:
It seems that every Group instance consists of only a single `Go()` block that starts the server and eventually waits on its error status.

Instead of using `errgroup.Group` for this, which I feel is overkill, could we just execute `runTestServer` in a goroutine and have it assign the error status to a channel? Then, read from that channel at the end of the test.

Reply:
Fair enough :)